README.md
[![Build Status](https://travis-ci.org/cpanic/WebService-Shippo.svg?branch=master)](https://travis-ci.org/cpanic/WebService-Shippo) [![Coverage Status](https://coveralls.io/repos/cpanic/WebService-Shippo/badge.svg?branch=master&service=github)](https://coveralls.io/github/cpanic/WebService-Shippo?branch=master)

## Shippo API Perl Client

The Shippo API can be used to automate and customize shipping capabilities for your e-commerce store or marketplace, enabling you to retrieve shipping rates, create and purchase shipping labels, track packages, and much more.

Shippo connects you with multiple shipping providers (USPS, UPS and Fedex, for example) through one interface, offering you great discounts on a selection of shipping rates.

You can sign-up for an account at https://goshippo.com/.

While Shippo do offer official API clients for a bevy of major languages, the venerable Perl is not among them. This client is a community offering that attempts to fill that void.

###PRE-RELEASE SOFTWARE

**Though functional, this software is still very much in the process of being
documented and should therefore be considered a work in progress.** 

### Version

* 0.0.21 (2015-11-14) on CPAN

### Requirements

Perl 5.8.8 minimum. 

Build tests have been conducted successfully on all later versions of Perl.

### Dependencies

<table>
<thead>
<tr>
<th align="left">Implementation</th>
<th align="left">Testing</th>
</tr>
</thead>
<tbody>
<tr>
<td valign="top">
<tt>boolean</tt><br/>
<tt>Clone</tt><br/>
<tt>File::HomeDir</tt><br/>
<tt>JSON::XS</tt><br/>
<tt>LWP</tt><br/>
<tt>LWP::Protocol::https</tt><br/>
<tt>MRO::Compat</tt><br/>
<tt>Params::Callbacks</tt><br/>
<tt>Path::Class</tt><br/>
<tt>Sub::Util</tt><br/>
<tt>Time::HiRes</tt><br/>
<tt>URI::Encode</tt><br/>
<tt>YAML::XS</tt><br/>
</td>
<td valign="top">
<tt>Data::Dumper::Concise</tt><br/>
<tt>Test::Deep</tt><br/>
</td>
</tr>
</tbody>
</table>

The Shippo API Perl Client depends on the non-core modules listed in the table above.

### Testing

Testing is standard operating procedure when installing Perl modules, since
test suites must normally complete successfully before a distribution can
be installed.

Be aware that, if you attempt to install this distribution without
first taking steps to configure Shippo API authentication, the testing
phase of installation will be _skipped_ rather than fail. It may still fail
if there are more fundamental problems with your Perl environment, but in
all likelihood the installation will complete without running all of its
tests.

You should probably run those tests.

Before installing the distribution, you should be in possession of a set
of Shippo API Authentication Tokens. You can get these by registering for
an account&mdash;go to https://goshippo.com/register. Once you have your
tokens, simply define the `SHIPPO_TOKEN` environment variable using the
**Private Auth Token** as the value. 

For example:

```shell
export SHIPPO_TOKEN="1a2b3c4d5e6ff7e8d9c0b1a21a2b3c4d5e6ff7e8"
```

Now, if you install the module, the tests should run.

### Installation

The Shippo API Perl Client is distributed on CPAN:

* http://search.cpan.org/dist/WebService-Shippo/lib/WebService/Shippo.pm

It is installed like the majority of Perl modules. If you have never installed a Perl module before then I recommend using `cpanminus` because it's super easy!

If you have never used `cpanminus` before then you can install this package
by running one of the following commands:

```shell
sudo -s 'curl -L cpanmin.us | perl - WebService::Shippo'

# If you're developing under PerlBrew then you probably don't want
# to use sudo...

curl -L cpanmin.us | perl - WebService::Shippo
```

If you **have** used `cpanminus` before then one of the following commands
will do the job:

```shell
sudo -s cpanm WebService::Shippo

# If you're developing under PerlBrew then you probably don't want
# to use sudo...

cpanm WebService::Shippo
```

### Installing from GitHub

Clone this repository only if you need to make changes. The distribution is 
managed using `Dist::Zilla`, which will have dependency requirements of
its own.

### Using the Shippo API Perl Client

```perl
use strict;
use WebService::Shippo;

# If it hasn't already done outside of the script, you
# must set your API key...
Shippo->api_key( 'PASTE YOUR PRIVATE AUTH TOKEN HERE' )
    unless Shippo->api_key;

my $address = Shippo::Address->create(
    object_purpose => 'PURCHASE',
    name           => 'John Smith',
    street1        => '6512 Greene Rd.',
    street2        => '',
    company        => 'Initech',
    phone          => '+1 234 346 7333',
    city           => 'Woodridge',
    state          => 'IL',
    zip            => '60517',
    country        => 'US',
    email          => 'user@gmail.com',
    metadata       => 'Customer ID 123456'
);

print 'Success with Address 1 : ', $address
```

All being well, you should see something like the following output:

```
Success with Address 1 : {
   "city" : "Woodridge",
   "company" : "Initech",
   "country" : "US",
   "email" : "user@gmail.com",
   "ip" : null,
   "is_residential" : null,
   "messages" : [],
   "metadata" : "Customer ID 123456",
   "name" : "John Smith",
   "object_created" : "2015-10-16T16:14:16.296Z",
   "object_id" : "475bb05d72b74a08a1d44b40ac85d635",
   "object_owner" : "******@*********.***",
   "object_purpose" : "PURCHASE",
   "object_source" : "FULLY_ENTERED",
   "object_state" : "VALID",
   "object_updated" : "2015-10-16T16:14:16.296Z",
   "phone" : "0012343467333",
   "state" : "IL",
   "street1" : "6512 Greene Rd.",
   "street2" : "",
   "street_no" : "",
   "zip" : "60517"
}
```
For additional integration notes, visit the [project wiki pages](https://github.com/cpanic/WebService-Shippo/wiki).
### Full API documentation

* For API documentation, go to https://goshippo.com/docs/ 
* For API support, contact support@goshippo.com with any questions.

### ACKNOWLEDGEMENTS

Thanks to Ali Saheli and the wonderful folks at Shippo for assistance
rendered.

### COPYRIGHT

This software is copyright &copy; 2015 by Iain Campbell.

You may distribute this software under the terms of either the GNU General
Public License or the Artistic License, as specified in the Perl README
file.

### SUPPORT / WARRANTY

The Shippo API Perl Client is free Open Source software; _it comes without
warranty of any kind._



Changes
Revision history for WebService-Shippo
 
{{$NEXT}}

0.0.21    2015-11-14 23:19:22+00:00 Europe/London
    - The currency refresh in WebService::Shippo::Currency now fails
      quietly, falling back to the hard-coded snapshot.
    - Some tidying up and a little re-factoring
0.0.20    2015-11-12 19:02:44+00:00 Europe/London
    - De-clutter interface to main package
    - Clean up serialization a bit
    - The collection "count" method works as a static method
    - POD updates
0.0.19    2015-11-12 01:45:32+00:00 Europe/London
    - Didn't run tests before releasing 0.0.18 to CPAN.
    - Should have run tests before releasing 0.0.18 to CPAN.
    - Travis you're the best!
    - Clear out debug from tests causing one failure and fix a 
      serialization bug in another test.
    - Once more, with feeling...
0.0.18    2015-11-12 01:32:55+00:00 Europe/London
    - Some work on the iterator
    - A new collector added, which uses the iterator to gather matches
      and some very basic tests.
    - A slew of POD updates
0.0.17    2015-11-10 19:10:39+00:00 Europe/London
    - Much POD, so wow!
    - Separated Item and Collection class definitions to ease
      their documentation.
    - Locale::Codes and Locale::Currency both pretty hairy to use
      accross different versions of Perl. Have come up with a 
      different plan until one of thoses works consistently on 
      on Perl versions.
0.0.16    2015-11-09 22:32:19+00:00 Europe/London
    - Some of the ancillary modules that didn't create Shippo::
      namespace aliases now do so for the sake of consistency.
    - There were some symbols exported on request and via tag :all; 
      they're no longer exported at all because I'm not sure they'll
      be used anyway.
    - Begun to separate API entity classes from their collection classes
      to make documentation easier. I'll continue this in the next release.
    - POD updates.
0.0.15    2015-11-08 21:19:47+00:00 Europe/London
    - With the exception of "id" methods that didn't started with
      "object_" but were identical in function to AUTOLOADed methods
      that did have that prefix have now been removed.
    - AUTOLOADed methods with always return '' instead of undef.
    - Collection classes (always the plural form of the item class) now
      respond to the Creator, Fetcher, Updater interfaces.
    - POD updates.
0.0.14    2015-11-05 21:09:08+00:00 Europe/London
    - Some tidying up
    - Some documentation
    - A slow day
0.0.13    2015-11-03 20:06:01+00:00 Europe/London
    - Some POD stubs added.
0.0.12    2015-11-03 18:55:11+00:00 Europe/London
    - Address->validate no longer refreshes subject of method call if 
      that subject is an address. The returned object is a validated
      address, not an opportunity to enrich the subject.
    - to_json is pretty if also passed a true as second argument
    - AUTOLOAD methods defined as pure accessors, not as mutators.
    - All entities now implement an iterator using the iterator method
    - Fix buggy ObjectList page transitions. Results were not correctly
      blessed as items but as lists.
    - Lots of stuff that just didn't smell right either got fixed or
      re-factored.
0.0.11    2015-11-02 17:31:00+00:00 Europe/London
    - The CarrierAccount "active" and "production" mutators have been
      replaced with "activate", "deactivate", "enable_test_mode" and
      "enable_production_mode" methods.
    - The non-OO interface is history; seems a little pointless
    - We're working with JSON a lot and so too is the client (probably)
      so the boolean, true, false functions brought in by boolean pragma
      are exported by default. So is callback.
    - Some re-factoring.
    - Now beginning to work on POD.
0.0.10    2015-10-25 19:56:30+00:00 Europe/London
    - Basic Authorization added.
    - Numerous doc and test change.
    - Had to remove Try::Tiny because it was giving grief under 5.8.8
    - Still having grief with Locale::Currency which works on some Perls
      and not others. If I avoid the really poor public interface with the
      patchy support on earlier and later versions of Perl, and use the 
      private Locale::Codes interface, I only have problems on Perl 5.14.
      Fighting the urge to avoid using it altogether by eliminating the
      dependency and replacing it with something that works properly and
      consistently. Grumble, grumble. Hope it can be fixed. And tested. On
      lots and lots of different Perls. Because that's good. Right?
0.0.9     2015-10-24 23:41:44+01:00 Europe/London
    - A reasonable minimum set of test finished. There can always be more
      tests and there will be in the future. At 96.1% coverage (local), 93%
      (Github), I'm now going to get some documentation done.
0.0.8     2015-10-24 16:16:29+01:00 Europe/London
    - Interim release.
    - Entities with asynchronous work-flows stable.
    - More tests.
    - More refactoring.
0.0.7     2015-10-20 20:00:55+01:00 Europe/London
    - Interim release (no more until the 23rd when I return home).
    - More core tests added for Rate and Shipment entities.
    - Most of the important user-facing methods take blocking callbacks.
    - A metric ton of re-factoring.
    - And a partridge in a pear tree.
0.0.6     2015-10-19 23:34:42+01:00 Europe/London
    - Interim release (reasonably stable)
    - More tests added
    - Lots of re-factoring (particularly of test code)

0.0.5     2015-10-19 20:16:14+01:00 Europe/London
    - Interim release (reasonably stable).
    - More tests added.
    - Lots and lots of re-factoring.
    
0.0.4     2015-10-17 00:01:33+01:00 Europe/London
    - Interim release.

0.0.3     2015-10-16 17:55:27+01:00 Europe/London
    - Interim release.
    
0.0.2     2015-10-16 01:07:52+01:00 Europe/London
    - Interim release.
    - API appears functional. Cruddy developer tests will now be built-out
      into something more useful and robust. Naturally, documentation is

lib/WebService/Shippo.pm
use strict;
use warnings;

package WebService::Shippo;
our $VERSION = '0.0.21';
# ABSTRACT: A Shippo Perl API wrapper

require WebService::Shippo::Addresses;
require WebService::Shippo::CarrierAccounts;
require WebService::Shippo::CustomsDeclarations;
require WebService::Shippo::CustomsItems;
require WebService::Shippo::Manifests;
require WebService::Shippo::Parcels;
require WebService::Shippo::Rates;
require WebService::Shippo::Refunds;
require WebService::Shippo::Shipments;
require WebService::Shippo::Transactions;
require WebService::Shippo::Config;

use boolean ':all';
use Params::Callbacks ( 'callbacks', 'callback' );
use base ( 'Exporter' );

our @EXPORT = qw(
    true
    false
    boolean
    callback
);

sub import
{
    my ( $class ) = @_;
    WebService::Shippo::Config->config;
    WebService::Shippo::Resource->api_credentials(
        @ENV{ 'SHIPPO_USER', 'SHIPPO_PASS' } )
        if $ENV{SHIPPO_USER} && !$ENV{SHIPPO_TOKEN};
    WebService::Shippo::Resource->api_key( $ENV{SHIPPO_TOKEN} )
        if $ENV{SHIPPO_TOKEN};
    goto &Exporter::import;
}

BEGIN {
    no warnings 'once';
    *api_key         = *WebService::Shippo::Resource::api_key;
    *api_credentials = *WebService::Shippo::Resource::api_credentials;
    *Shippo::        = *WebService::Shippo::;
}

1;
__END__
=pod

=encoding utf8

=head1 UNDER CONSTRUCTION

B<Though functional, this software is still in the process of being
documented and should, therefore, be considered a work in progress.> 

=head1 NAME

WebService::Shippo - Shippo API Client

=head1 SYNOPIS

B<Note>: though scripts and modules must always C<use WebService::Shippo;>
to import the client software, the C<WebService::> portion of that package
namespace may be dropped when subsequently referring to the main package
or any of its resource classes. For example, C<WebService::Shippo::Address>
and C<Shippo::Address> refer to the same class. 

To compel the developer to continue using the C<WebService::> prefix does
seem like an unreasonable form of torture, besides which, it probably
doesn't leave much scope for indenting code as some class names would be
very long. Use it, or don't use it. It's entirely up to you.

    use strict;
    use LWP::UserAgent;
    use WebService::Shippo;
    
    # If you aren't using a config file or the environment (SHIPPO_TOKEN=...)
    # to supply your API key, you can do so here:
    
    Shippo->api_key('PASTE YOUR AUTH TOKEN HERE')
        unless Shippo->api_key;
    
    # Complete example illustrating the the process of Shipment creation
    # through to label generation.
    #
    # Create a Shipment object:
    
    my $shipment = Shippo::Shipment->create(
        object_purpose => 'PURCHASE',
        address_from   => {
            object_purpose => 'PURCHASE',
            name           => 'Shawn Ippotle',
            company        => 'Shippo',
            street1        => '215 Clayton St.',
            city           => 'San Francisco',
            state          => 'CA',
            zip            => '94117',
            country        => 'US',
            phone          => '+1 555 341 9393',
            email          => 'shippotle@goshippo.com'
        },
        address_to => {
            object_purpose => 'PURCHASE',
            name           => 'Mr Hippo',
            company        => '',
            street1        => 'Broadway 1',
            street2        => '',
            city           => 'New York',
            state          => 'NY',
            zip            => '10007',
            country        => 'US',
            phone          => '+1 555 341 9393',
            email          => 'mrhippo@goshippo.com'
        },
        parcel => {
            length        => '5',
            width         => '5',
            height        => '5',
            distance_unit => 'in',
            weight        => '2',
            mass_unit     => 'lb'
        }
    );
    
    # Retrieve shipping rates and select preferred rate:
    
    my $rates = Shippo::Shipment->get_shipping_rates( $shipment->object_id );
    my $preferred_rate = $rates->item(2);
    
    # Purchase label at the preferred rate:
    
    my $transaction = Shippo::Transaction->create(
        rate            => $preferred_rate->object_id,
        label_file_type => 'PNG',
    );
    
    # Get the shipping label:
    
    my $label_url = Shippo::Transaction->get_shipping_label( $transaction->object_id );
    my $browser   = LWP::UserAgent->new;
    $browser->get( $transaction->label_url, ':content_file' => './sample.png' );
    
    # Print the transaction object...
    
    print "Transaction:\n", $transaction;

    --[content dumped to console]--
    Transaction:
    {
       "commercial_invoice_url" : null,
       "customs_note" : "",
       "label_url" : "https://shippo-delivery-east.s3.amazonaws.com/da2e68fe85f94a9ebca458d9f9d
    2446b.PNG?Signature=BjD2JMQt0ATd5jUWAKm%2B6FHcBPM%3D&Expires=1477323662&AWSAccessKeyId=AKIA
    JGLCC5MYLLWIG42A",
       "messages" : [],
       "metadata" : "",
       "notification_email_from" : false,
       "notification_email_other" : "",
       "notification_email_to" : false,
       "object_created" : "2015-10-25T15:41:01.182Z",
       "object_id" : "da2e68fe85f94a9ebca458d9f9d2446b",
       "object_owner" : "******@*********.***",
       "object_state" : "VALID",
       "object_status" : "SUCCESS",
       "object_updated" : "2015-10-25T15:41:02.494Z",
       "order" : null,
       "pickup_date" : null,
       "rate" : "3c76e81733d7417b9a801ce957f4219d",
       "submission_note" : "",
       "tracking_history" : [],
       "tracking_number" : "9499907123456123456781",
       "tracking_status" : {
          "object_created" : "2015-10-25T15:41:02.451Z",
          "object_id" : "02ce6dbd6d5a48cfb764fdeb0cb6e404",
          "object_updated" : "2015-10-25T15:41:02.451Z",
          "status" : "UNKNOWN",
          "status_date" : null,
          "status_details" : ""
       },
       "tracking_url_provider" : "https://tools.usps.com/go/TrackConfirmAction_input?origTrackN
    um=9499907123456123456781",
       "was_test" : true
    }
    --[end of content]--

The sample code in this synopsis produced the following label:

=over 2

=item * L<https://github.com/cpanic/WebService-Shippo/blob/master/sample.png>

=back

=head1 DESCRIPTION

Shippo connects you with multiple shipping providers (USPS, UPS and Fedex,
for example) through one interface, offering you great discounts on a
selection of shipping rates. You can sign-up for an account at 
L<https://goshippo.com/>.

The Shippo API can be used to automate and customize shipping capabilities
for your e-commerce store or marketplace, enabling you to retrieve shipping 
rates, create and purchase shipping labels, track packages, and much more.

Though Shippo I<do> offer official API clients for a bevy of major languages, 
the venerable Perl 5 was not included in that list. This community offering 
attempts to correct that omission ;-)

=head2 API Resources

Access to all Shippo API resources is via URLs relative to the same encrypted
API endpoint (https://api.goshippo.com/v1/). The client provides object and
collection classes to help with the nitty-gritty of dealing with those
resources:

=over 2

=item * L<Addresses|WebService::Shippo::Address>

=item * L<Parcels|WebService::Shippo::Parcel>

=item * L<Shipments|WebService::Shippo::Shipment>

=item * L<Rates|WebService::Shippo::Rate>

=item * L<Transactions|WebService::Shippo::Transaction>

=item * L<Customs Items|WebService::Shippo::CustomsItem>

=item * L<Customs Declarations|WebService::Shippo::CustomsDeclaration>

=item * L<Refunds|WebService::Shippo::Refund>

=item * L<Manifests|WebService::Shippo::Manifest>

=item * L<Carrier Accounts|WebService::Shippo::CarrierAccount>

=back

Each item class has a related collection class with a similar name I<in the
plural form>. The rationale behind this is that the Shippo API can be used
to retrieve single objects with the C<fetch> method, and collections of
objects with the C<all> method, and different behaviours may be applied
to collections, which is why both forms exist.

=head2 Request & Response Data

The Perl client ensures that requests are properly encoded and passed to the
correct API endpoints using appropriate HTTP methods. There is documentation
for each API resource, containing more details on the values accepted by and
returned for a given resource (see L<FULL API DOCUMENTATION>).

All API requests return responses encoded as JSON strings, which the client
converts into Perl blessed object references of the correct type. As a rule,
any resource attribute documented in the API specification will have an
accessor of the same same in a Perl instance of that object.

=head2 REST & Disposable Objects

The Shippo API is built with simplicity and RESTful principles in mind:
B<POST> requests are used to create objects, B<GET> requests to fetch and
list objects, and B<PUT> requests to update objects. The Perl client provides
C<create>, C<fetch>, C<all> and C<update> methods for use with resource
objects that permit such operations.

Addresses, Parcels, Shipments, Rates, Transactions, Refunds, Customs Items and
Customs Declarations are disposable objects. This means that once you create
an object, you cannot change it. Instead, create a new one with the desired
values. Carrier Accounts are the exception and may be updated via B<PUT>
requests.

=head1 OBJECTS vs COLLECTIONS

Whereas a GET request to L<https://api.goshippo.com/v1/addresses/d799c2679e644279b59fe661ac8fa488>
might produce a single address whose structure would resemble this:

    {
       "object_state":"VALID",
       "object_purpose":"PURCHASE",
       "object_source":"FULLY_ENTERED",
       "object_created":"2014-07-09T02:19:13.174Z",
       "object_updated":"2014-07-09T02:19:13.174Z",
       "object_id":"d799c2679e644279b59fe661ac8fa488",
       "object_owner":"shippotle@goshippo.com",
       "name":"Shawn Ippotle",
       "street1":"215 Clayton St.",
       "street2":"",
       "city":"San Francisco",
       "state":"CA",
       "zip":"94117",
       "country":"US",
       "phone":"15553419393",
       "email":"shippotle@goshippo.com",
       "ip":null,
       "is_residential":true,
       "messages":[],
       "metadata":"Customer ID 123456"
    }

A GET request to L<https://api.goshippo.com/v1/addresses/> might produce a
collection of address objects whose structure would resemble this:

    {
       "count":3,
       "next":null,
       "previous":null,
       "results":[
          {
             "object_state":"VALID",
             "object_purpose":"PURCHASE",
             "object_source":"FULLY_ENTERED",
             "object_created":"2014-07-16T23:20:31.089Z",
             "object_updated":"2014-07-16T23:20:31.089Z",
             "object_id":"747207de2ba64443b645d08388d0309c",
             "object_owner":"shippotle@goshippo.com",
             "name":"Shawn Ippotle",
             "street1":"215 Clayton St.",
             "street2":"",
             "city":"San Francisco",
             "state":"CA",
             "zip":"94117",
             "country":"US",
             "phone":"15553419393",
             "email":"shippotle@goshippo.com",
             "is_residential":true,
             "ip":null,
             "messages":[
    
             ],
             "metadata":"Customer ID 123456"
          },
          {...},
          {...}
       ]
    }

The Shippo API does not distinguish between a single object and a collection
of objects. It deals purely with JSON strings and both object and collection
are notionally considered to be the same type of object.

Nevertheless, the Perl client B<does> distinguish between a single object and
a collection of objects. For each API resource, the client defines an I<Item
Class> (e.g. C<L<WebService::Shippo::Address>>) and a corresponding
I<Collection Class> (e.g. C<L<WebService::Shippo::Addresses>>). The name of
the collection class is always the plural form of the item class. Furthermore,
any objects making up the C<results> of a collection will be blessed as 
instances of the appropriate item class.

Both item and collection classes implement different interfaces, but they
also exhibit common behaviours. Collection classes will respond to methods
that aid item access and cursor navigation, while both types of class respond
(where appropriate) to C<create>, C<update>, C<fetch>, C<all>, C<iterate> and
C<collect> methods.

=head1 METHODS

=head2 api_key

Get or set the key used by API requests for Shippo's token-based
authentication. This is Shippo's preferred method of authentication.

=over 2

=item * Return the token currently being used for authentication.

    my $api_key = Shippo->api_key;

=item * Set the token to be used for authentication.
    
    Shippo->api_key($auth_token);

The C<api_key> method is chainable when used as a setter.

=back

=head2 api_credentials

Get or set the login credentials used by API requests for Shippo's legacy
authentication. Legacy authentication means encoding the HTTP Authorization
header for Basic Authentication so, even though requests and repsonses are
encrypted, you should still consider using the token-based authentication
instead (see C<api_key>). 

=over 2

=item * Return the login credentials currently being used for authentication.

    my ($username, $password) = Shippo->api_credentials;

=item * Set the credentials to be used for authentication.

    Shippo->api_credentials($username, $password);

The C<api_credentials> method is chainable when used as a setter.

=back

Whenever a configuration specifies both token and login credentials, the
client will always favour token-based authentication. If C<api_key> and
C<api_credentials> are both set manually then it is the most recently
set mechanism that defines the HTTP Authorization header.

=head1 EXPORTS

The C<WebService::Shippo> package exports a number of helpful subroutines by
default:

=head2 true

    my $fedex_account = Shippo::CarrierAccount->create(
        carrier    => 'fedex',
        account_id => '<YOUR-FEDEX-ACCOUNT-NUMBER>',
        parameters => { meter => '<YOUR-FEDEX-METER-NUMBER>' },
        test       => true,
        active     => true
    );

Returns a scalar value which will evaluate to true. 

Since the I<lingua franca> connecting Shippo's API and the Perl client is
JSON, it can feel more natural to think in those terms. Thus, C<true> may be
used in place of C<1>. Now, when creating a new object from a JSON example,
any literal and accidental use of C<true> or C<false> is much less likely
to result in misery.

See Ingy's L<boolean> package for more guidance.

=head2 false

    my $fedex_account = Shippo::CarrierAccount->create(
        carrier    => 'fedex',
        account_id => '<YOUR-FEDEX-ACCOUNT-NUMBER>',
        parameters => { meter => '<YOUR-FEDEX-METER-NUMBER>' },
        test       => false,
        active     => false
    );

Returns a scalar value which will evaluate to false. 

Since the I<lingua franca> connecting Shippo's API and the Perl client is
JSON, it can feel more natural to think in those terms. Thus, C<false> may be
used in place of C<0>. Now, when creating a new object from a JSON example,
any literal and accidental use of C<true> or C<false> is much less likely
to result in misery.

See Ingy's L<boolean> package for more guidance.

=head2 boolean

    my $bool = boolean($value);

Casts a scalar value to a boolean value (C<true> or C<false>).

See Ingy's L<boolean> package for more guidance.

=head2 callback

    $put_all_accounts_in_test_mode = Shippo::CarrierAccounts->collect( 
        callback { shift->enable_test_mode } );
    $put_all_accounts_in_test_mode->();
    
Many of the Perl client's methods allow the use of lambda functions for
block operations and filtration.

See L<Params::Callbacks> for more guidance.

=head1 CONFIGURATION

While the client does provide C<api_key> and C<api_credentials> methods to
help with authentication, hard-coding such calls in anything more mission 
critical than a simple test script may I<not> be the best way to go.

As soon as it is imported, one of the first things the client does is search
a number of locations for a L<YAML-encoded|https://en.wikipedia.org/wiki/YAML>
configuration file. The first one it finds is loaded.

In order, the locations searched are as follows:

=over 2

=item * C<./.shipporc>

=item * C</I<path>/I<to>/I<home>/.shipporc>

=item * C</I<etc>/shipporc>

=item * C</I<path>/I<to>/I<perl>/I<module>/I<install>/I<lib>/WebService/Shippo/Config.yml>

=back

The configuration file is very simple and needs to have the following
structure, though not all elements are mandatory:

    ---
    username: martymcfly@pinheads.org
    password: yadayada
    private_token: f0e1d2c3b4a5968778695a4b3c2d1e0f96877869
    public_token: 96877869f0e1d2c3b4a5968778695a4b3c2d1e0f
    default_token: private_token

At a minimum, your configuration should define values for C<private_token> and
C<public_token>. These are your Shippo Private and Publishable Auth tokens,
which are found on your L<Shippo API page|https://goshippo.com/user/apikeys/>.

=head1 FULL API DOCUMENTATION

=over 2

=item * For API documentation, go to L<https://goshippo.com/docs/>

=item * For API support, contact L<support@goshippo.com|mailto:support@goshippo.com> with any 
questions.

=back

=head1 REPOSITORY

=over 2

=item * L<https://github.com/cpanic/WebService-Shippo>

=back

=head1 AUTHOR

Iain Campbell <cpanic@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright E<copy> 2015 by Iain Campbell.

You may distribute this software under the terms of either the GNU General
Public License or the Artistic License, as specified in the Perl README
file.


=cut
